packages/vscode-extension/src/utilities/extensionContext.ts

```diff
@@ -15,15 +19,187 @@ export const extensionContext = new Proxy<ExtensionContext>({} as ExtensionConte
   },
 });
 
-let _appRootFolder: string | null = null;
+export class AppRootFolder {
+  private onChangeAppRootListeners: Array<(newAppRoot: string) => void> = [];
 
-export function setAppRootFolder(appRootFolder: string) {
-  _appRootFolder = appRootFolder;
+  constructor(private appRoot: string) {}
+
+  onChangeAppRoot(listener: (newAppRoot: string) => void): Disposable {
+    this.onChangeAppRootListeners.push(listener);
+    return {
+      dispose: () => {
+        const index = this.onChangeAppRootListeners.indexOf(listener);
+        if (index > -1) {
+          this.onChangeAppRootListeners.splice(index, 1);
+        }
+      },
+    };
+  }
+
+  getAppRoot(): string {
+    if (!this.appRoot) {
+      throw new Error("App root not set.");
+    }
+    return this.appRoot;
+  }
+
+  setAppRoot(newAppRoot: string): void {
+    this.appRoot = newAppRoot;
+    this.onChangeAppRootListeners.forEach((listener) => {
+      listener(newAppRoot);
+    });
+    Logger.debug(`App root was set to: ${this.appRoot}.`);
+  }
 }
 
-export function getAppRootFolder() {
-  if (!_appRootFolder) {
-    throw new Error("App root folder has not been set");
+export const getCurrentLaunchConfig = (): LaunchConfigurationOptions => {
+  const launchConfiguration = workspace.getConfiguration(
+    "launch",
+    workspace.workspaceFolders![0].uri
+  );
+
+  const configurations = launchConfiguration.get<Array<Record<string, any>>>("configurations")!;
+
+  const RNIDEConfiguration = configurations.find(
+    ({ type }) => type === "react-native-ide" || type === "radon-ide" // for compatibility we want to support old configuration type name
+  );
+
+  if (!RNIDEConfiguration) {
+    return {};
   }
-  return _appRootFolder;
+
+  const { android, appRoot, ios, isExpo, metroConfigPath, env } = RNIDEConfiguration;
+
+  return { android, appRoot, ios, isExpo, metroConfigPath, env };
+};
+
+export function findAppRootCandidates(maxSearchDepth: number = 3): string[] {
+  const candidates: string[] = [];
+  const searchedFileNames = [
+    "metro.config.js",
+    "metro.config.ts",
+    "app.json",
+    "app.config.js",
+    "app.config.ts",
+  ];
+
+  // In order to optimize the search time we exclude directories,
+  // that shouldn't contain applications.
+  const excludedDirectoryPatterns: RegExp[] = [/^node_modules$/, /^ios$/, /^android$/, /^\..+/];
+
+  const searchQueue: [string, number][] | undefined = workspace.workspaceFolders?.map(
+    (workspaceFolder) => {
+      return [workspaceFolder.uri.path, 0];
+    }
+  );
+
+  if (searchQueue === undefined) {
+    Logger.warn("[FindFiles] Could not determine active workspace");
+    return [];
+  }
```

Comment:
I think it'd make more sense to instead early return on `workspace.workspaceFolders` being undefined.
```suggestion
  if (workspace.workspaceFolders === undefined) {
    Logger.warn("[FindFiles] Could not determine active workspace");
    return [];
  }

  const searchQueue: [string, number][] = workspace.workspaceFolders.map(
    (workspaceFolder) => {
      return [workspaceFolder.uri.path, 0];
    }
  );
```